src/github.com/getlantern/errlog/errlog.go

```diff
@@ -0,0 +1,540 @@
+	// TODO: attach systemInfo in aggregator/reporter instead, as it remains
+	// the same in the lifetime of a program.
+	*systemInfo
+	*ProxyingInfo
+	*UserLocale
+	*UserAgentInfo
+}
+
+func (e *Error) String() string {
+	var buf bytes.Buffer
+	_, _ = buf.WriteString(e.Desc)
+	if e.Op != "" {
+		_, _ = buf.WriteString(" op=" + e.Op)
+	}
+	if e.systemInfo != nil {
+		_, _ = buf.WriteString(e.systemInfo.String())
+	}
+	if e.ProxyingInfo != nil {
+		_, _ = buf.WriteString(e.ProxyingInfo.String())
+	}
+	if e.UserLocale != nil {
+		_, _ = buf.WriteString(e.UserLocale.String())
+	}
+	if e.UserAgentInfo != nil {
+		_, _ = buf.WriteString(e.UserAgentInfo.String())
+	}
+	for k, v := range e.Extra {
+		_, _ = buf.WriteString(" " + k + "=" + v)
+	}
+	return buf.String()
+}
+
+// Customized marshaller to marshal extra fields to same level as other struct fields
+/*func (e Error) MarshalJSON() ([]byte, error) {
+	var buf bytes.Buffer
+	// safe to ignore return value as error returned is always nil
+	_, _ = buf.WriteString(fmt.Sprintf(`{"package":"%s","type":"%s","desc":"%s"`, e.GoPackage, e.GoType, e.Desc))
+	if e.Extra != nil && len(e.Extra) > 0 {
+		_, _ = buf.WriteString(",")
+		for k, v := range e.Extra {
+			_, _ = buf.WriteString(fmt.Sprintf(`"%s":"%s"`, k, v))
+		}
+	}
+	_, _ = buf.WriteString("}")
+	return buf.Bytes(), nil
+}*/
+
+type ErrorLogger struct {
+	goPackage string
+	logger    golog.Logger
+	*systemInfo
+}
+
+type withFunc func(e *Error)
+
+func WithOp(op string) withFunc {
+	return func(e *Error) {
+		e.Op = op
+	}
+}
+
+func WithProxy(info *ProxyingInfo) withFunc {
+	return func(e *Error) {
+		e.ProxyingInfo = info
+	}
+}
+
+func WithLocale() withFunc {
+	return func(e *Error) {
+		lang, _ := jibber_jabber.DetectLanguage()
+		country, _ := jibber_jabber.DetectTerritory()
+		e.UserLocale = &UserLocale{
+			time.Now().Format("MST"),
+			lang,
+			country,
+		}
+	}
+}
+
+func WithUserAgent(ua string) withFunc {
+	return func(e *Error) {
+		e.UserAgentInfo = &UserAgentInfo{ua}
+	}
+}
+
+func WithField(k, v string) withFunc {
+	return func(e *Error) {
+		e.Extra[k] = v
+	}
+}
+
+func (c *ErrorLogger) Log(err error, with ...withFunc) {
+	errOp, goType, desc, extra := parseError(err)
+	actual := &Error{
+		GoPackage:  c.goPackage,
+		GoType:     goType,
+		Desc:       desc,
+		Op:         errOp,
+		Extra:      extra,
+		systemInfo: c.systemInfo,
+	}
+	for _, f := range with {
+		f(actual)
+	}
+	currentReporter.Report(actual)
+	c.logger.Error(actual.String())
+}
+
+func ErrorLoggerFor(goPackage string) *ErrorLogger {
+	version, _ := osversion.GetHumanReadable()
+	return &ErrorLogger{
+		goPackage: goPackage,
+		logger:    golog.LoggerFor(goPackage),
+		systemInfo: &systemInfo{
+			OSType:    runtime.GOOS,
+			OSArch:    runtime.GOARCH,
+			OSVersion: version,
+		},
+	}
+}
+
+type Reporter interface {
+	Report(*Error)
+}
+
+var currentReporter Reporter = &StdReporter{}
+
+func ReportTo(r Reporter) {
+	currentReporter = r
```

Comment:
This is where I assume I'll have to plug my event processing mechanism, correct @fffw ?